### DISCUSSION TITLE
Running flows in local server

### DISCUSSION THREAD

**Discussion by carloszanella** [PrefectHQ/prefect#5163] [Q&A] [answered]
Hey everyone,

I am having issues when trying to use Prefect Server locally. I have Prefect deployed in a kubernetes cluster and this is working fine, but what I would like to do is to test/inspect/debug my flows and tasks on prefect server locally. 

Before going any further, this might or not be relevant: I am running in an M1 chip. I had to tweak the docker-compose config to use a different GraphQL image because when trying to `prefect server start` this container wouldn't start. So I am not running the server with prefect cli but using `docker-compose` instead.

I have all my 6 containers up and running. I can access the dashboard in the browser. I created a tenant.
```
CONTAINER ID   IMAGE                               COMMAND                  CREATED          STATUS                    PORTS                            NAMES
530246312bdd   prefecthq/ui:core-0.14.12           "/docker-entrypoint.…"   44 minutes ago   Up 44 minutes (healthy)   80/tcp, 0.0.0.0:8080->8080/tcp   local-prefect_ui_1
e1a9c23ad1ae   prefecthq/apollo:core-0.14.12       "tini -g -- bash -c …"   44 minutes ago   Up 44 minutes (healthy)   0.0.0.0:4200->4200/tcp           local-prefect_apollo_1
8ed559783e31   prefecthq/server:core-0.14.12       "tini -g -- python s…"   44 minutes ago   Up 44 minutes                                              local-prefect_towel_1
1718cd7f7dba   prefecthq/server:core-0.14.12       "tini -g -- bash -c …"   44 minutes ago   Up 44 minutes (healthy)   0.0.0.0:4201->4201/tcp           local-prefect_graphql_1
eca12984c6ac   fedormelexin/graphql-engine-arm64   "graphql-engine serve"   44 minutes ago   Up 44 minutes (healthy)   0.0.0.0:3000->3000/tcp           local-prefect_hasura_1
14c5daa3788f   postgres:11                         "docker-entrypoint.s…"   44 minutes ago   Up 44 minutes (healthy)   0.0.0.0:5432->5432/tcp           local-prefect_postgres_1

```

What I would like to do is to test the same code that would deploy my flow to the Kubernetes cluster, but do it in this local deployment of the server. The problem is I haven't managed to get any flow run to show up on the dashboard.

I am using `Local` storage and a `LocalAgent`. Using the docker alternative (DockerRun, DockerStorage, DockerAgent) would require me building images to test my code which is not nice. I start an agent with the proper labels and I have a very simple flow and the agent sees it:
```
[2021-11-20 19:36:19,850] INFO - agent | Found 1 flow run(s) to submit for execution.
[2021-11-20 19:36:20,029] INFO - agent | Deploying flow run 0f6be811-729d-4973-9d6f-02e23acb8ea4
```

From the client, I can see the run executed successfully, but nothing shows up in the UI/Dashboard. If I start the run from the dashboard itself it also doesn't register the run there. **What is the proper way to do this?** Where can I see the task logs? If my flow breaks I have no idea how to debug it.
```
client.get_flow_run_state(flow_run_id)
<Success: "All reference tasks succeeded.">
```

**Comment by zanieb** [comment]
Hi!

We run Prefect Server on M1 macs using the built-in docker-compose command without alteration. What issue did you run into there? I'm presuming your issue here is that you're using a different graphql image. Did you use `prefect server config` to generate your compose file or did you make it from scratch? Is the UI indicating that it is connecting to the apollo url?

Just as a note: You can actually use a DockerAgent without using DockerStorage or building images. The default image is just the latest prefect image and you can provide an environment variable `EXTRA_PIP_PACKAGES` to install requirements https://docs.prefect.io/orchestration/flow_config/docker.html#installing-extra-dependencies-at-runtime.

**Comment by carloszanella** [comment]
hi @madkinsz  thanks for the reply. 

> Did you use prefect server config to generate your compose file or did you make it from scratch?

I used `prefect server config`. I tried initially starting the server via CLI. Then I noticed that GraphQL image wasn't starting. I checked the error and I found a discussion in hasura/grahpql that the image isn't supported on M1. In this discussion they offered an alternative image that does work. I changed the docker-compose file to use this image and prefect server started fine.

> We run Prefect Server on M1 macs using the built-in docker-compose command without alteration

Could it be a prefect version thing? I am using `0.14.12`. The GraphQL image I am using is: `fedormelexin/graphql-engine-arm64`. 

> Is the UI indicating that it is connecting to the apollo url?

How can I verify that? The UI looks to be working mostly fine but the runs don't show. I can see the registered flows, the agents, the labels, the projects.

**Reply by zanieb** [reply]
> I used prefect server config. I tried initially starting the server via CLI. Then I noticed that GraphQL image wasn't starting. I checked the error and I found a discussion in hasura/grahpql that the image isn't supported on M1. In this discussion they offered an alternative image that does work. I changed the docker-compose file to use this image and prefect server started fine.

Interesting. Can you share the error? I wonder why we've never encountered it.

> Could it be a prefect version thing? I am using 0.14.12. The GraphQL image I am using is: fedormelexin/graphql-engine-arm64.

Possibly, but I'm doubtful. Using the latest version may be worth a try.

> How can I verify that? The UI looks to be working mostly fine but the runs don't show. I can see the registered flows, the agents, the labels, the projects.

Hm. If you can see the rest it should be connecting fine. There is a connection indicator in the upper right that should show where you are connecting to. Can you confirm that the environment variables in your flow run's subprocess contain the same connection string for the API as the UI is using? 

If you go to the interactive API in the UI and run the query

```graphql

query { flow_run { id start_time name } }

```

are there results?

> client.get_flow_run_state(flow_run_id)
> <Success: "All reference tasks succeeded.">

Where are you instantiating this client?

**Reply by carloszanella** [reply]
> Interesting. Can you share the error? I wonder why we've never encountered it.

The graphql container just keeps restarting. When trying to run it individually I get: 
```WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested```

You can check here that the image is not supported on M1:
https://github.com/hasura/graphql-engine/issues/6337

What graphql image do you use?

> are there results?

Yes I can see the run with the query

> Where are you instantiating this client?

Locally, inside a unit test. Same place where I create/register the flow and trigger a run on it

**Comment by zanieb** [comment, marked as answer]
> What graphql image do you use?

We're using the graphql-engine image, I'm just running `prefect server start` -- do you have enough memory allocated #4611 https://github.com/hasura/graphql-engine/issues/6337#issuecomment-969300069 ?

> Yes I can see the run with the query

So the UI can see the runs via that query but they are not appearing in your dashboard or the page for the flow?

**Reply by carloszanella** [reply]
hey @madkinsz thanks. I increased memory to 8GB on docker and it looks stable now, Prefect UI seem to be working well too - I can see all the information there.
